File: algorithms/game_of_life/game_of_life.py
# Conway's Game of Life - Norvig-inspired implementation

def neighbors(cell):
    """Return the 8 neighbors of a cell."""
    x, y = cell
    return {(x+dx, y+dy) for dx in [-1, 0, 1] for dy in [-1, 0, 1] if dx or dy}


def step(cells):
    """Compute next generation of live cells."""
    candidates = cells | {n for cell in cells for n in neighbors(cell)}
    return {cell for cell in candidates if (cell in cells and neighbor_count(cell, cells) in (2, 3)) or 
            (cell not in cells and neighbor_count(cell, cells) == 3)}


def neighbor_count(cell, cells):
    """Count live neighbors of a cell."""
    return sum(n in cells for n in neighbors(cell))


def life(cells, generations):
    """Run Game of Life for given number of generations."""
    for _ in range(generations):
        cells = step(cells)
    return cells

File: algorithms/game_of_life/game_of_life.rs
// Conway's Game of Life - Rust implementation
use std::collections::HashSet;

type Cell = (i32, i32);

fn neighbors(cell: Cell) -> Vec<Cell> {
    let (x, y) = cell;
    let mut result = Vec::new();
    for dx in -1..=1 {
        for dy in -1..=1 {
            if dx != 0 || dy != 0 {
                result.push((x + dx, y + dy));
            }
        }
    }
    result
}

fn neighbor_count(cell: Cell, cells: &HashSet<Cell>) -> usize {
    neighbors(cell).iter().filter(|n| cells.contains(n)).count()
}

pub fn step(cells: &HashSet<Cell>) -> HashSet<Cell> {
    let mut candidates = HashSet::new();
    
    // Add all live cells and their neighbors
    for &cell in cells {
        candidates.insert(cell);
        for neighbor in neighbors(cell) {
            candidates.insert(neighbor);
        }
    }
    
    candidates
        .into_iter()
        .filter(|&cell| {
            let count = neighbor_count(cell, cells);
            (cells.contains(&cell) && (count == 2 || count == 3)) || 
            (!cells.contains(&cell) && count == 3)
        })
        .collect()
}

pub fn life(mut cells: HashSet<Cell>, generations: usize) -> HashSet<Cell> {
    for _ in 0..generations {
        cells = step(&cells);
    }
    cells
}
